src/ikernel.ts

```diff
@@ -404,6 +417,17 @@ namespace IKernel {
      * The done handler for the kernel future.
      */
     onDone: () => void;
+
+    /**
+     * Register hook for IOPub messages.
+     * 
+     * @returns a disposable to unregister the hook.
+     * 
+     * #### Notes
+     * The most recently-registered hook is run first.
+     */
+    registerMessageHook(hook: (msg: KernelMessage.IIOPubMessage) => boolean): void;
```

Comment:
I see, what you are saying, by returning a disposable delegate with a closure, we are keeping a reference in memory to that object.  From an API standpoint, I think it would make more sense to align the methods then so that you use `register/remove` on both the kernel and the future.